File: javascript/easy/pathSum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {boolean}
 */
var hasPathSum = function(root, sum) {
    return pathSum(root, 0, sum);
};

var pathSum = function(root, running, sum) {
    if(root == null) {
        return false;
    } else if(root.left == null && root.right == null) {
        return ((running + root.val) == sum);
    } else {
        return pathSum(root.left, (running + root.val), sum) || pathSum(root.right, (running + root.val), sum);
    }
};

File: ruby/easy/path_sum.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} sum
# @return {Boolean}
def has_path_sum(root, sum)
    return false unless root
    queue = []
    queue.unshift({ node: root, sum: 0 })
    while(!queue.empty?)
        current = queue.pop()
        node = current[:node]
        running = (node.val + current[:sum])
        return true if (running == sum) && (node.left.nil? && node.right.nil?)
        queue.unshift({ node: node.left, sum: running }) if node.left
        queue.unshift({ node: node.right, sum: running }) if node.right
    end
    return false
end
